bitcoin/BRWalletManager.c

```diff
@@ -690,17 +699,79 @@ BRWalletManagerGetWallet (BRWalletManager manager) {
 
 extern void
 BRWalletManagerConnect (BRWalletManager manager) {
+    pthread_mutex_lock (&manager->lock);
     BRSyncManagerConnect (manager->syncManager);
+    pthread_mutex_unlock (&manager->lock);
 }
 
 extern void
 BRWalletManagerDisconnect (BRWalletManager manager) {
+    pthread_mutex_lock (&manager->lock);
     BRSyncManagerDisconnect (manager->syncManager);
+    pthread_mutex_unlock (&manager->lock);
 }
 
 extern void
 BRWalletManagerScan (BRWalletManager manager) {
+    pthread_mutex_lock (&manager->lock);
     BRSyncManagerScan (manager->syncManager);
+    pthread_mutex_unlock (&manager->lock);
+}
+
+extern void
+BRWalletManagerSetMode (BRWalletManager manager, BRSyncMode mode) {
+    pthread_mutex_lock (&manager->lock);
+    if (mode != manager->mode) {
+        // get the currently known block height
+        uint64_t blockHeight = BRSyncManagerGetBlockHeight (manager->syncManager);
+
+        // kill the sync manager to prevent any additional callbacks from occuring
+        BRSyncManagerDisconnect (manager->syncManager);
+        BRSyncManagerFree (manager->syncManager);
+
+        // load up the blocks/peers saved to disk
+        BRArrayOf(BRMerkleBlock*) blocks = initialBlocksLoad(manager);
+        BRArrayOf(BRPeer) peers = initialPeersLoad(manager);
```

Comment:
These are surely written to the persistent store by now?

Reply:
As best can be reasoned, yes.